### PR TITLE
Specifics strategy

### PR DESCRIPTION
**Pull request by d-michail**
This PR continues the discussion at #459 by introducing a `GraphSpecificsStrategy` interface, which is provided to the graph on construction time. The strategy object contains all factories that are needed in order to build all internal data structures (maps, etc.).

I have also cleaned up all places where default values were used, and delegated all calls to the strategy. Moreover, I fixed the issues where the fast lookup specifics were not using the same edge factory as the base specifics class.

After this one, we can go forward and add a separate `opt` package with specialized graph implementations. It should be straightforward to replace all maps, etc.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/graph/specifics/FastLookupDirectedSpecifics.java [hunk 1]
```diff
@@ -105,11 +133,12 @@ public FastLookupDirectedSpecifics(
...
     public E getEdge(V sourceVertex, V targetVertex)
     {
-        List<E> edges = touchingVerticesToEdgeMap.get(new Pair<>(sourceVertex, targetVertex));
+        Set<E> edges = touchingVerticesToEdgeMap.get(new Pair<>(sourceVertex, targetVertex));
         if (edges == null || edges.isEmpty())
             return null;
-        else
-            return edges.get(0);
+        else { 
+            return edges.stream().findFirst().orElse(null);
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/specifics/FastLookupDirectedSpecifics.java]: Is there an advantage to this over edges.iterator().next()?  There's no need for the option check since we already tested for isEmpty above.

File: jgrapht-core/src/main/java/org/jgrapht/graph/specifics/FastLookupUndirectedSpecifics.java [hunk 2]
```diff
@@ -106,12 +133,12 @@ public FastLookupUndirectedSpecifics(
...
     public E getEdge(V sourceVertex, V targetVertex)
     {
-        List<E> edges =
+        Set<E> edges =
             touchingVerticesToEdgeMap.get(new UnorderedPair<>(sourceVertex, targetVertex));
         if (edges == null || edges.isEmpty())
             return null;
         else
-            return edges.get(0);
+            return edges.stream().findFirst().orElse(null);
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/specifics/FastLookupUndirectedSpecifics.java]: Ditto

File: jgrapht-core/src/main/java/org/jgrapht/graph/specifics/UndirectedSpecifics.java [hunk 3]
```diff
@@ -43,46 +44,51 @@
 {
     private static final long serialVersionUID = 6494588405178655873L;
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/specifics/UndirectedSpecifics.java]: I think this has to be updated.

File: jgrapht-core/src/main/java/org/jgrapht/graph/AbstractBaseGraph.java [hunk 4]
```diff
@@ -287,15 +326,18 @@ public Object clone()
 
             newGraph.vertexSupplier = this.vertexSupplier;
             newGraph.edgeSupplier = this.edgeSupplier;
-
+            newGraph.type = type;
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/AbstractBaseGraph.java]: Oops, nice catch :)

File: jgrapht-core/src/main/java/org/jgrapht/graph/specifics/DirectedSpecifics.java [hunk 5]
```diff
@@ -43,46 +44,51 @@
 {
     private static final long serialVersionUID = 8971725103718958232L;
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/specifics/DirectedSpecifics.java]: I think this needs to be updated.

File: jgrapht-core/src/main/java/org/jgrapht/graph/GraphSpecificsStrategy.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    BiFunction<Graph<V, E>, GraphType, Specifics<V, E>> getSpecificsFactory();
+
+    /**
+     * Get a supplier of maps with a predictable iteration order.
+     * 
+     * @return a supplier of maps with a predictable iteration order.
+     * @param <K1> the key type
+     * @param <V1> the value type 
+     */
+    default <K1, V1> Supplier<Map<K1, V1>> getPredictableOrderIterationMapFactory()
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/GraphSpecificsStrategy.java]: Let's shorten this to getPredictableOrderMapFactory?

**Comment by jkinable** [comment]
@jsichi I'll go over it in the next few days.

**Comment by jsichi** [comment]
I'm gonna go ahead and merge now...if Joris sees issues later, we can open up a followup.
